File: mobile/ios/SacredAttempt/NavigationPolicy.swift
import Foundation

enum NavigationPolicy {
    static let host = "upsc-cse-tracker-adarsh.vercel.app"
    static let home = URL(string: "https://\(host)/dashboard")!

    static func isHTTPS(_ url: URL) -> Bool {
        url.scheme?.lowercased() == "https" && url.host != nil && url.user == nil
            && url.password == nil && (url.port == nil || url.port == 443)
    }
    static func isInternal(_ url: URL) -> Bool {
        isHTTPS(url) && url.host?.lowercased() == host
    }
}

File: mobile/android/app/src/main/java/com/adarsh/upsc/NavigationPolicy.java
package com.adarsh.upsc;

import java.net.URI;

/** Only the owner's exact HTTPS origin may remain inside the authenticated WebView. */
public final class NavigationPolicy {
    public static final String HOST = "upsc-cse-tracker-adarsh.vercel.app";
    public static final String HOME = "https://" + HOST + "/dashboard";
    public static boolean isHttps(String value) {
        try {
            URI uri = URI.create(value);
            return "https".equalsIgnoreCase(uri.getScheme()) && uri.getHost() != null
                && uri.getRawUserInfo() == null && (uri.getPort() == -1 || uri.getPort() == 443);
        } catch (IllegalArgumentException error) { return false; }
    }
    public static boolean isInternal(String value) {
        return isHttps(value) && HOST.equalsIgnoreCase(URI.create(value).getHost());
    }
    private NavigationPolicy() {}
}
